src/EventStore.Core/Services/RedactionService.cs

```diff
@@ -0,0 +1,244 @@
+			_readIndex = readIndex;
+			_switchChunksLock = switchChunksLock;
+		}
+
+		public void Handle(RedactionMessage.GetEventPosition message) {
+			ThreadPool.QueueUserWorkItem(_ => {
+				try {
+					GetEventPosition(message.EventStreamId, message.EventNumber, message.Envelope);
+				} catch (Exception ex) {
+					Log.Error(ex, "An error has occurred when getting position for stream: {stream}, event number: {eventNumber}.",
+						message.EventStreamId, message.EventNumber);
+					message.Envelope.ReplyWith(
+						new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.UnexpectedError, Array.Empty<EventPosition>()));
+				}
+			});
+		}
+
+		private void GetEventPosition(string streamName, long eventNumber, IEnvelope envelope) {
+			var streamId = _readIndex.GetStreamId(streamName);
+			var result = _readIndex.ReadEventInfo_KeepDuplicates(streamId, eventNumber);
+
+			var eventPositions = new EventPosition[result.EventInfos.Length];
+
+			for (int i = 0; i < result.EventInfos.Length; i++) {
+				var eventInfo = result.EventInfos[i];
+				var logPos = eventInfo.LogPosition;
+				var chunk = _db.Manager.GetChunkFor(logPos);
+				var localPosition = chunk.ChunkHeader.GetLocalLogPosition(logPos);
+				var chunkEventOffset = chunk.GetActualRawPosition(localPosition);
+
+				// all the events returned by ReadEventInfo_KeepDuplicates() must exist in the log
+				// since the log record was read from the chunk to check for hash collisions.
+				if (chunkEventOffset < 0)
+					throw new Exception($"Failed to fetch actual raw position for event at log position: {logPos}");
+
+				if (chunkEventOffset > uint.MaxValue)
+					throw new Exception($"Actual raw position for event at log position: {logPos} is larger than uint.MaxValue: {chunkEventOffset}");
+
+				eventPositions[i] = new EventPosition(
+					logPosition: logPos,
+					chunkFile: Path.GetFileName(chunk.FileName),
+					chunkVersion: chunk.ChunkHeader.Version,
+					chunkComplete: chunk.ChunkFooter is { IsCompleted: true },
+					chunkEventOffset: (uint) chunkEventOffset);
+			}
+
+			envelope.ReplyWith(
+				new RedactionMessage.GetEventPositionCompleted(GetEventPositionResult.Success, eventPositions));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkLock message) {
+			if (_switchChunksLock.TryAcquire())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkLockCompleted(SwitchChunkLockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunkUnlock message) {
+			if (_switchChunksLock.TryRelease())
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Success));
+			else
+				message.Envelope.ReplyWith(
+					new RedactionMessage.SwitchChunkUnlockCompleted(SwitchChunkUnlockResult.Failed));
+		}
+
+		public void Handle(RedactionMessage.SwitchChunk message) {
```

Comment:
it would probably be good if the service ensured that the lock was held by whomever is sending the request.

perhaps the grpc call should generate an id (guid) and include it in lock/switch/unlock messages. the redaction service could then check that the lock is held and by the right person when a switchchunk message is received, and it might help in the grpccall itself to be able to 'finally unlock' safely even if the lock was not successfully acquired

Reply:
good idea :+1: